### PR TITLE
Combine necessary file for edge route in size calculation

### PR DESCRIPTION
**Pull request by eps1lon**
Just including `page.js` in the size calculation isn't enough since it may require other chunks.

In a Webpack world we'd use the Webpack stats instead for accurate calculations without having to read from the filesystem. Turbopack doesn't have these available so we just read every file listed in the the `files` entry for that specific route in the middleware manifest.

I know not enough about all these manifest to be able to generalize this so for now this only works for the edge-ssr calculation which regressed in https://github.com/vercel/next.js/pull/64798 but wasn't caught since the significant bundle increase was in a different chunk.

Before:
Screenshot 2024-04-26 at 17 41 40
-- https://github.com/vercel/next.js/pull/64798#issuecomment-2067124032

After:
Screenshot 2024-04-26 at 17 41 09
-- https://github.com/vercel/next.js/pull/65053#issuecomment-2077890361

Closes NEXT-3228

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7eb04f8a8ae7867646a3a08b4b23f47d454ecac6

`TURBOPACK=1 pnpm test-start test/e2e/app-dir/app-external/app-external.test.ts` (turbopack)
- app dir - external dependency > should be able to opt-out 3rd party packages being bundled in server components

<details>
<summary>Expand output</summary>

  ● app dir - external dependency › should be able to opt-out 3rd party packages being bundled in server components

    expect(received).toContain(expected) // indexOf

    Expected substring: "opt-out-react-version: 18.2.0"
    Received string:    "<!DOCTYPE html><html><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width, initial-scale=1\"/><link rel=\"preload\" as=\"script\" fetchPriority=\"low\" href=\"/_next/static/chunks/node_modules__pnpm_e15454._.js\"/><script src=\"/_next/static/chunks/4b5e9_next_dist_esm_build_templates_app-page_6eb009.js\" async=\"\"></script><script src=\"/_next/static/chunks/_639177._.js\" async=\"\"></script><script src=\"/_next/static/chunks/_ef26de._.js\" async=\"\"></script><meta name=\"next-size-adjust\"/></head><body><div>Server: index.default<br/>Server subpath: subpath.default<br/>Client: index.default<br/>Client subpath: subpath.default<br/><p id=\"optout-react-version\">opt-out-react-version: 18.3.0</p></div><script src=\"/_next/static/chunks/node_modules__pnpm_e15454._.js\" async=\"\"></script><script>(self.__next_f=self.__next_f||[]).push([0]);self.__next_f.push([2,null])</script><script>self.__next_f.push([1,\"1:I[\\\"[project]/node_modules/.pnpm/next@file+..+next-repo-4cde3ce31c239f05efc3ebe8c9e846b4abcfd26f9f9bd507485b265e052c57f1+packa_p2z6iaz563f3ulxwdmfwhanivu/node_modules/next/dist/client/components/app-router.js [app-client] (ecmascript)\\\",[\\\"static/chunks/_639177._.js\\\"],\\\"default\\\"]\\n3:I[\\\"[project]/app/react-server/optout/client.js [app-client] (ecmascript)\\\",[\\\"static/chunks/_639177._.js\\\",\\\"static/chunks/_ef26de._.js\\\"],\\\"default\\\"]\\n4:I[\\\"[project]/node_modules/.pnpm/next@file+..+next-repo-4cde3ce31c239f05efc3ebe8c9e846b4abcfd26f9f9bd507485b265e052c57f1+packa_p2z6iaz563f3ulxwdmfwhanivu/node_modules/next/dist/client/components/layout-router.js [app-client] (ecmascript)\\\",[\\\"static/chunks/_639177._.js\\\"],\\\"default\\\"]\\n5:I[\\\"[project]/node_modules/.pnpm/next@file+..+next-repo-4cde3ce31c239f05efc3ebe8c9e846b4abcfd26f9f9bd507485b265e052c57f1+packa_p2z6iaz563f3ulxwdmfwhanivu/node_modules/next/dist/client/components/render-from-template-context.js [app-client] (ecmascript)\\\",[\\\"static/chunks/_639177._.js\\\"],\\\"default\\\"]\\n7:I[\\\"[project]/node_modules/.pnpm/next@file+..+next-repo-4cde3ce31c239f05efc3ebe8c9e846b4abcfd26f9f9bd507485b265e052c57f1+packa_p2z6iaz563f3ulxwdmfwhanivu/node_modules/next/dist/client/components/error-boundary.js [app-client] (ecmascript)\\\",[\\\"static/chunks/_639177._.js\\\"],\\\"default\\\"]\\n8:[]\\n\"])</script><script>self.__next_f.push([1,\"0:[[[\\\"$\\\",\\\"script\\\",\\\"script-0\\\",{\\\"src\\\":\\\"/_next/static/chunks/_639177._.js\\\",\\\"async\\\":true}]],[\\\"$\\\",\\\"$L1\\\",null,{\\\"buildId\\\":\\\"i468OU4Llc_AC-lsKmsWK\\\",\\\"assetPrefix\\\":\\\"\\\",\\\"initialCanonicalUrl\\\":\\\"/react-server/optout\\\",\\\"initialTree\\\":[\\\"\\\",{\\\"children\\\":[\\\"react-server\\\",{\\\"children\\\":[\\\"optout\\\",{\\\"children\\\":[\\\"__PAGE__\\\",{}]}]}]},\\\"$undefined\\\",\\\"$undefined\\\",true],\\\"initialSeedData\\\":[\\\"\\\",{\\\"children\\\":[\\\"react-server\\\",{\\\"children\\\":[\\\"optout\\\",{\\\"children\\\":[\\\"__PAGE__\\\",{},[[\\\"$L2\\\",[\\\"$\\\",\\\"div\\\",null,{\\\"children\\\":[\\\"Server: index.default\\\",[\\\"$\\\",\\\"br\\\",null,{}],\\\"Server subpath: subpath.default\\\",[\\\"$\\\",\\\"br\\\",null,{}],[\\\"$\\\",\\\"$L3\\\",null,{}],[\\\"$\\\",\\\"p\\\",null,{\\\"id\\\":\\\"optout-react-version\\\",\\\"children\\\":\\\"opt-out-react-version: 18.3.0\\\"}]]}]],null],null]},[\\\"$\\\",\\\"$L4\\\",null,{\\\"parallelRouterKey\\\":\\\"children\\\",\\\"segmentPath\\\":[\\\"children\\\",\\\"react-server\\\",\\\"children\\\",\\\"optout\\\",\\\"children\\\"],\\\"error\\\":\\\"$undefined\\\",\\\"errorStyles\\\":\\\"$undefined\\\",\\\"errorScripts\\\":\\\"$undefined\\\",\\\"template\\\":[\\\"$\\\",\\\"$L5\\\",null,{}],\\\"templateStyles\\\":\\\"$undefined\\\",\\\"templateScripts\\\":\\\"$undefined\\\",\\\"notFound\\\":\\\"$undefined\\\",\\\"notFoundStyles\\\":\\\"$undefined\\\",\\\"styles\\\":[[\\\"$\\\",\\\"script\\\",\\\"script-0\\\",{\\\"src\\\":\\\"/_next/static/chunks/_ef26de._.js\\\",\\\"async\\\":true}]]}],null]},[\\\"$\\\",\\\"$L4\\\",null,{\\\"parallelRouterKey\\\":\\\"children\\\",\\\"segmentPath\\\":[\\\"children\\\",\\\"react-server\\\",\\\"children\\\"],\\\"error\\\":\\\"$undefined\\\",\\\"errorStyles\\\":\\\"$undefined\\\",\\\"errorScripts\\\":\\\"$undefined\\\",\\\"template\\\":[\\\"$\\\",\\\"$L5\\\",null,{}],\\\"templateStyles\\\":\\\"$undefined\\\",\\\"templateScripts\\\":\\\"$undefined\\\",\\\"notFound\\\":\\\"$undefined\\\",\\\"notFoundStyles\\\":\\\"$undefined\\\",\\\"styles\\\":null}],null]},[[\\\"$\\\",\\\"html\\\",null,{\\\"children\\\":[[\\\"$\\\",\\\"head\\\",null,{}],[\\\"$\\\",\\\"body\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"$L4\\\",null,{\\\"parallelRouterKey\\\":\\\"children\\\",\\\"segmentPath\\\":[\\\"children\\\"],\\\"error\\\":\\\"$undefined\\\",\\\"errorStyles\\\":\\\"$undefined\\\",\\\"errorScripts\\\":\\\"$undefined\\\",\\\"template\\\":[\\\"$\\\",\\\"$L5\\\",null,{}],\\\"templateStyles\\\":\\\"$undefined\\\",\\\"templateScripts\\\":\\\"$undefined\\\",\\\"notFound\\\":[[\\\"$\\\",\\\"title\\\",null,{\\\"children\\\":\\\"404: This page could not be found.\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"fontFamily\\\":\\\"system-ui,\\\\\\\"Segoe UI\\\\\\\",Roboto,Helvetica,Arial,sans-serif,\\\\\\\"Apple Color Emoji\\\\\\\",\\\\\\\"Segoe UI Emoji\\\\\\\"\\\",\\\"height\\\":\\\"100vh\\\",\\\"textAlign\\\":\\\"center\\\",\\\"display\\\":\\\"flex\\\",\\\"flexDirection\\\":\\\"column\\\",\\\"alignItems\\\":\\\"center\\\",\\\"justifyContent\\\":\\\"center\\\"},\\\"children\\\":[\\\"$\\\",\\\"div\\\",null,{\\\"children\\\":[[\\\"$\\\",\\\"style\\\",null,{\\\"dangerouslySetInnerHTML\\\":{\\\"__html\\\":\\\"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}\\\"}}],[\\\"$\\\",\\\"h1\\\",null,{\\\"className\\\":\\\"next-error-h1\\\",\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\",\\\"margin\\\":\\\"0 20px 0 0\\\",\\\"padding\\\":\\\"0 23px 0 0\\\",\\\"fontSize\\\":24,\\\"fontWeight\\\":500,\\\"verticalAlign\\\":\\\"top\\\",\\\"lineHeight\\\":\\\"49px\\\"},\\\"children\\\":\\\"404\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\"},\\\"children\\\":[\\\"$\\\",\\\"h2\\\",null,{\\\"style\\\":{\\\"fontSize\\\":14,\\\"fontWeight\\\":400,\\\"lineHeight\\\":\\\"49px\\\",\\\"margin\\\":0},\\\"children\\\":\\\"This page could not be found.\\\"}]}]]}]}]],\\\"notFoundStyles\\\":[],\\\"styles\\\":null}]}]]}],null],null],\\\"couldBeIntercepted\\\":false,\\\"initialHead\\\":[false,\\\"$L6\\\"],\\\"globalErrorComponent\\\":\\\"$7\\\",\\\"missingSlots\\\":\\\"$W8\\\"}]]\\n\"])</script><script>self.__next_f.push([1,\"6:[[\\\"$\\\",\\\"meta\\\",\\\"0\\\",{\\\"name\\\":\\\"viewport\\\",\\\"content\\\":\\\"width=device-width, initial-scale=1\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"1\\\",{\\\"charSet\\\":\\\"utf-8\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"2\\\",{\\\"name\\\":\\\"next-size-adjust\\\"}]]\\n2:null\\n\"])</script></body></html>"

      47 |       expect(result).toContain('Client: index.default')
      48 |       expect(result).toContain('Client subpath: subpath.default')
    > 49 |       expect(result).toContain('opt-out-react-version: 18.2.0')
         |                      ^
      50 |     })
      51 |   })
      52 |

      at toContain (e2e/app-dir/app-external/app-external.test.ts:49:22)
      at Object.<anonymous> (e2e/app-dir/app-external/app-external.test.ts:43:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/edge-route-size-comparison | Change |
| - | - | - | - |
| buildDuration | 17.4s | 14.5s | N/A |
| buildDurationCached | 8s | 7.6s | N/A |
| nodeModulesSize | 359 MB | 359 MB | ✓ |
| nextStartRea..uration (ms) | 410ms | 392ms | N/A |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 7eb04f8a8ae7867646a3a08b4b23f47d454ecac6

File: test/.stats-app/stats-config.js [hunk 1]
```diff
@@ -31,6 +34,39 @@ const clientGlobs = [
...
+        const manifestJson = await fs.readFile(
+          path.join(nextAppDir, '.next/server/middleware-manifest.json')
+        )
+        const manifest = JSON.parse(manifestJson)
+        console.log({ manifest, functions: manifest.functions })
+        const manifestFileEntry = path.relative(
+          path.join(nextAppDir, '.next'),
+          path.join(nextAppDir, fileName)
+        )
+        console.log({ manifestFileEntry })
```
Comment (huozhi) [inline test/.stats-app/stats-config.js:48]: rm the log?
```suggestion
```

**Review by huozhi — APPROVED** [review]

File: test/.stats-app/stats-config.js (hunk 1 above)
Comment (eps1lon) [reply test/.stats-app/stats-config.js]: Yep, thanks for catching this 👍🏻

**Pull request merged**